src/components/date/date.jsx

```diff
@@ -0,0 +1,264 @@
+import React, {
+  forwardRef,
+  useEffect,
+  useLayoutEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+} from 'react';
+import calendarSvg from '../../svgs/calendar.svg';
+import cautionSvg from '../../svgs/caution.svg';
+import Calendar from '../calendar/calendar.jsx';
+import FormControl from '../form-control/form-control.jsx';
+import IconButton from '../icon-button/icon-button.jsx';
+import Icon from '../icon/icon.jsx';
+import useDropdownClose from '../../hooks/use-dropdown-close.js';
+import styles from './date.css';
+
+/* eslint-disable react/prop-types */
+function ControlIcons(props) {
+  return (
+    <div className={styles.icons}>
+      {!!props.validationMessage && (
+        <Icon
+          className={styles['invalid-icon']}
+          icon={cautionSvg}
+          label={props.validationMessage}
+        />
+      )}
+      {props.readOnly ? (
+        <Icon
+          icon={calendarSvg}
+          title={props.label}
+          label={`${props.label} calendar icon`}
+        />
+      ) : (
+        <IconButton
+          onPress={props.onPress}
+          icon={calendarSvg}
+          title={props.label}
+          label={`${props.label} calendar button`}
+        />
+      )}
+    </div>
+  );
+}
+/* eslint-enable react/prop-types */
+
+function toDateStringFormat(date) {
+  return date ? date.toLocaleDateString(undefined, { month: 'short', year: 'numeric', day: 'numeric' }) : '';
+}
+
+function toFullDateFormat(date) {
+  return date ? date.toISOString().slice(0, 10) : undefined;
+}
+
+function fromFullDateFormat(string) {
+  const date = new window.Date(`${string}T00:00:00`);
+  if (date.toDateString() === 'Invalid Date') {
+    return undefined;
+  }
+
+  return date;
+}
+
+const Date = forwardRef(function Date(props, forwardedRef) {
+  const backupRef = useRef();
+  const ref = forwardedRef || backupRef;
+  const containerRef = useRef();
+  const inputRef = useRef();
+  const [active, setActive] = useState(false);
+  const [editing, setEditing] = useState(!!props.validationMessage);
+  const [expanded, setExpanded] = useState(false);
+  const [validationMessage, setValidationMessage] = useState(props.validationMessage);
+  const [value, setValue] = useState(fromFullDateFormat(props.value));
+  const classNames = {
+    input: validationMessage ? styles['invalid-input'] : styles.input,
+    ...props.classNames,
+    layout: {
+      container: styles.container,
+      calendar: styles['calendar-container'],
+      ...props.classNames.layout,
+    },
+  };
+  const ids = {
+    input: props.id,
+    label: `${props.id}-label`,
+    validationMessage: `${props.id}Hint`,
+  };
+
+  function onBlur(event) {
+    if (containerRef.current.contains(event.relatedTarget)) return;
+
+    // Given the validation spec, any provided validation messages are cleared on blur.
+    // If the native node is still invalid then re-render with the native validation messages.
+    inputRef.current.setCustomValidity('');
```

Comment:
This tiny bit can be extracted to be the new validation spec.

Reply:
Are we not using `useValidationMessage` anymore?

Reply:
At the moment, no. I think the existing hook does too much? I want to try to implement MDS Input validation by replacing the existing hook with these new `useEffect` and `useLayoutEffect` and the `onBlur` callback. But that's another story to avoid too much code changes.